Dependencies/Tests/Deprecation/Validation/Chore/Typos

monheimx9 opened the pull request:
Hi Yatekii,
This is my first time contributing to an open source project besides my own, I'm probably missing some organisation and planning :^)
Since we started working on it almost at the same time by coincidence with @jacg I also suggest that you could take a look if the direction I'm taking is aligning with what you want too
At the same times I was also working on the clippy warnings too, there are commits with the same code

I started to write some unit testing too, and also began to write integration tests, I'm planning to make more of them in the future

I'm working on a CLI tool that use that library to replace the python script that I previously made, I'll be using your library in the small company were I work

On the lib.rs file I think it's not necessary to call twice node::Text as follow:
```rust  
Text::new("")
.add(svg::node::Text::new(line))
```
This should be sufficient
```rust
Text::new(line)
```

While exploring the Todo's I also made that change, is it what you intended?

```rust
  // TODO: validate QR IBAN / QRID matches.
        match &options.reference {
            Reference::Qrr(x) => x.validate_qriid(&options.account)?,
            Reference::Scor(s) => s.validate_iid(&options.account)?,
            Reference::None => {}
        }
```

In the Esr library I took the liberty of deprecating the "try_new()" function and replacing it with two new functions, one with checksum, the other without.

Do you think this was a good idea?
I wanted to give the user the possibility of generating a valid Esr number directly from the library.
I'm thinking about doing the same for the SCOR reference in the future

Anyway, feel free to point out any changes that you want
I wish you a good day

monheimx9 committed:
Bump edition to 2021

monheimx9 committed:
Bump svg to 0.17.0

Text::new() without args is deprecated
It requires now the content to be passed as an argument
I've updated the nodes so that now all the chained method are after the
first new() call, and it doesn't need to be in a .add() method

monheimx9 committed:
Resolve chrono::Date deprecation

Using NaiveDate instead of Date
I don't know why Utc was needed, I should write a few test cases to see
if there's any weird behaviours triggered with that change

monheimx9 committed:
Bump other libs

No conflict to resolve at this time, further testing might be needed

monheimx9 committed:
Clippy Lint

Statisfying unneeded let statement

monheimx9 committed:
Clippy Lints

I must do as saint clippy says

monheimx9 committed:
Issue with reference placement

The reference shown to the right side of the QRCode was misplaced,
causing some overlap on the left side of the QRCode
The reference was using the x value of the Receipt text block

monheimx9 committed:
ESR Invalid format error message

The error message shown was valid for iso11649 and not ESR

monheimx9 committed:
checksum return type changed to Result

It is not really a necessary change but it made testing a bit more
effective
(And also clippy :^) on the return statement)

monheimx9 committed:
ESR Tests writing

Few tests written for the ESR module

monheimx9 committed:
Revert back to #6beb2970

Probably will use &mut Group in the future to avoid returning a copy of
the same variable

monheimx9 committed:
Bump packages versions

monheimx9 committed:
New functions to instatiate ESR

Deprecating try_new()
Now choosing if the checksum is provided directly or not
Also new func to test if the checksum is valid

monheimx9 committed:
Writing test and changing to_string() impl

Using the Display trait instead of the ToString to satisfy clippy
+ tests

monheimx9 committed:
Validating QR-IID and IID with ESR+SCOR

Tests required
Now it will test if ESR match QR-IID and if SCOR match non-QR-IID

monheimx9 committed:
Display instead of ToString traits

monheimx9 committed:
Adding Debug trait

monheimx9 committed:
Writing integration tests

monheimx9 committed:
Merge branch 'master' into dev-update-crates

monheimx9 committed:
test.pdf should also be ignored

monheimx9 committed:
Rust formatter is complaining

jacg commented:
> I started to write some unit testing too, and also began to write integration tests, I'm planning to make more of them in the future

I have some (very ugly, but working in principle) tests for checking that the data encoded in the QR code are as expected:

1. Verify output of `QRBill::qr_data()`
1. Roundtrip:
   - write QR code to PNG file
   - read file
   - decode using [`bardecoder`](https://github.com/piderman314/bardecoder) 
1. Same as the previous one but keeping the image in memory without writing to file. This one doesn't work yet.

They use [`rstest`](https://docs.rs/rstest/latest/rstest/) to create an example fixture which is used in the three aforementioned varieties of test. Eventually I'd like to use [`proptest`](https://proptest-rs.github.io/proptest/proptest/index.html) to generate multiple examples. (@Yatekii how would you feel about adding these to dev-dependencies?)

Let me know if you are doing anything along these lines, so we can avoid duplicating effort, or share ideas.

Incidentally, it was pretty easy to find examples which produced test failures when going via the PNG file. I haven't had the time to figure out which component in the cycle was responsible for the failure, so it might just be that one of the dev-dependencies used only in the test is the weak like, in which case the failures are spurious, but maybe there's a more fundamental problem with the real code itself.

Yatekii commented on src/esr.rs:
```diff
@@ -7,58 +15,156 @@ pub struct Esr {
     #[error("Checksum is invalid.")]
     InvalidChecksum,
+    #[error("The following IBAN is not a QR-IID")]
+    InvalidQriid,
```
Can you add the IIBAN as a field and use it in the error message?

monheimx9 replied:
Something like this ? -> https://github.com/Yatekii/qrbill/pull/12/commits/c4e83b56319ca1700b82f7c0ecfdc49f13417324

Yatekii replied:
Yep :)

Yatekii commented on src/esr.rs:
```diff
@@ -7,58 +15,156 @@ pub struct Esr {
+        Ok(Self { number })
+    }
+
+    /// Instantiate a new [`Esr`] struct
```
```suggestion
    /// Instantiate a new [`Esr`] struct
    ///
```

Add an empty line to actually separate the headline from the second line.

Yatekii commented on src/esr.rs:
````diff
@@ -7,58 +15,156 @@ pub struct Esr {
+    }
+
+    /// Instantiate a new [`Esr`] struct
+    /// The checksum should already be present at the end of the string!
+    /// If your reference doesn't have the checksum calculated use this method instead
+    /// ```
+    /// qrbill::esr::Esr::try_without_checksum("00000024072049".to_string());
+    /// ```
````
```suggestion
    /// The checksum should already be present at the end of the string!
    ///
    /// If your reference doesn't have the checksum calculated use [`Esr::try_without_checksum`] instead.
```

monheimx9 committed:
IBAN value in Error Message

monheimx9 committed:
Better function commenting

Co-authored-by: Noah Hüsser <noah@huesser.dev>

monheimx9 committed:
Better function commenting

Co-authored-by: Noah Hüsser <noah@huesser.dev>

monheimx9 commented:
> I have some (very ugly, but working in principle) tests for checking that the data encoded in the QR code are as expected:
> 
> 1. Verify output of `QRBill::qr_data()`
> 2. Roundtrip:
>    
>    * write QR code to PNG file
>    * read file
>    * decode using [`bardecoder`](https://github.com/piderman314/bardecoder)
> 3. Same as the previous one but keeping the image in memory without writing to file. This one doesn't work yet.
> 

@jacg Great idea, I've only worked a little bit with "Image Buffer" and don't feel very confident in it, if you're cooking something I would gladly see it

> They use [`rstest`](https://docs.rs/rstest/latest/rstest/) to create an example fixture which is used in the three aforementioned varieties of test. Eventually I'd like to use [`proptest`](https://proptest-rs.github.io/proptest/proptest/index.html) to generate multiple examples. (Yatekii how would you feel about adding these to dev-dependencies?)
> 
> Let me know if you are doing anything along these lines, so we can avoid duplicating effort, or share ideas.

Sure, at this time I'm working on a separate CLI tool and I'm building a validator/builder for Swico S1 Syntax for the extra_info field
See https://github.com/Yatekii/qrbill/issues/13
For now it's only WIP in my local repo, but depending on how it goes I could add it here instead


If @Yatekii  is okay using adding rstest and proptest I would gladly change my current unit tests and learn how to use rstest and proptest

monheimx9 committed:
IBAN value in Error Message (forgot in test)

Yatekii commented:
Sure, feel free to use all the amazing crates :) We should also set up some CI pipeline :)

monheimx9 committed:
Breaking changes/Basic Exemple/Integration tests

Introducing a breaking change on the StructuredAddress struc
Fields were publics while we had a new() function
This made possible to avoid error detection on field's data

Also made pub use of NaiveDate and CountryCode so they can be used
outside the library

Integration tests are far from complete, still figuring out how to make
clean tests without too much boilerplate

Public use of dependencies

monheimx9 committed:
rstest dev-dep/fix logic of qriid-iid validation

As discussed previously, here's now the rstest integration
While doing so I've noticed that I made a mistake in the qriid/iid
validation
It made possible to enter a valid qriid while not providing any qrr
reference
So I changed the way I did the check previously and ported the errors up
in the main lib, while also adding a new trait to Iban to determine if
the iban is qr or iid

Basic testing

monheimx9 commented:
> Sure, feel free to use all the amazing crates :) We should also set up some CI pipeline :)

Then here are my commits :^)

It fixes a mistake I did on the qriid validation match on the qr reference, also added the rstest dev-dependency
You will see that I made public use of NaiveDate and CountryCode to avoid to import them again in any crate that depend on qrbill

I've introduced "kind of" a breaking change, since you made a new() function on the StructuredAddress struc, I've made the fields private to force error detection on creation

Also, how do we deal with this typo ? ^_^
```rust
#[derive(Debug)]
pub enum Address {
    Cobined(CombinedAddress),
    Structured(StructuredAddress),
}
```
Do we keep Cobined instead of Combined :D ?

If everything is ok for you I think this pull request is long enough
I should have squashed a few commits and should have rebased before pushing my branch, I'm sorry I'm still on the beginner side on the Git etiquette

My next PR's will be cleaner

Regarding the CI pipeline I've never done any, but if I can be of any help feel free to ask I also think it could be nice to add one on this project

jacg commented:
Sorry about my radio silence. Hopefully I'll find some time to contribute a bit in the next few days.

GHA OK for the CI?

Yatekii commented:
Thanks a lot folks!

re: Typo: I would just fix it :) I don't think this crate sees much usage apart from the three of us and I rather fix an API that is an internal detail (and not on the wire) than keep it broken forever :)

re: CI: yeah GHA is great :) Maybe I find some time tomorrow :)

monheimx9 committed:
Typo

Cobined -> Combined

Yatekii commented:
Sorry this went to rot :( I put all of my OSS work on hold. Wanna pick this up again and get it in? :)

monheimx9 commented:
Hi Yatekii :^)
The good news is that in the meantime I was able to test it in production for a while (along with the other pull request with the Swico syntax)

We've generated arround 2'000 invoices without any issue 👍 

It might not be the place but I also took the opportunity to start to learn embedded Rust on my esp32 boards and I also use probe-rs, so, thank you a lot for that !! It's an amazing project

Since @jacg made a lot of work I need to learn how to properly resolve conflicts
I will come back during this winter to address this correctly, along with my other PR

In the meantime, wish you all the best !
